Support python slicing with tensor inputs.

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #165074

when the slice is tensor, we decompose it to .item() call and pass the unbacked symbol to the slice to avoid DDE. 
the diff also fix an existing bug in codegen_dynamic_slice_size in the cpp wrapper.  a +1 should be -1 making it match 
python codegen.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

laithsakka committed:
WIP Support python slicing with data depedennt inptu tensors maybe

[ghstack-poisoned]

laithsakka committed:
Update on "[WIP] Support python slicing with tensor inputs."


allow things like
``` 
#!/usr/bin/env python
import torch

print("="*60)
print("Testing tensor slicing with torch.compile")
print("="*60)

# Test 1: Simple eager mode
print("\n1. Eager mode test:")
x = torch.randn(10)
idx = torch.tensor(4)
result = x[:idx]
print(f"   x[:idx] where idx=4: result.shape = {result.shape}")
assert result.shape[0] == 4
print("   ✓ Eager mode works!")

# Test 2: With torch.compile
print("\n2. Compiled mode test:")
def slice_fn(x, idx):
    return x[:idx]

try:
    compiled_fn = torch.compile(slice_fn)
    x = torch.randn(10)
    idx = torch.tensor(4)
    result = compiled_fn(x, idx)
    print(f"   Compiled x[:idx] where idx=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Compiled mode works!")
except Exception as e:
    print(f"   ✗ Compiled mode failed: {e}")
    import traceback
    traceback.print_exc()

# Test 3: With dynamic slicing from sum
print("\n3. Dynamic slicing with sum:")
def dynamic_slice_fn(x, lengths):
    idx = lengths.sum()
    return x[:idx]

try:
    compiled_fn = torch.compile(dynamic_slice_fn)
    x = torch.randn(10)
    lengths = torch.tensor([1, 1, 1, 1])
    result = compiled_fn(x, lengths)
    print(f"   Compiled x[:lengths.sum()] where sum=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Dynamic slicing works!")
except Exception as e:
    print(f"   ✗ Dynamic slicing failed: {e}")
    import traceback
    traceback.print_exc()

print("\n" + "="*60)
print("SUMMARY: Check results above")
print("="*60)

```


[ghstack-poisoned]

laithsakka committed:
Update on "[WIP] Support python slicing with tensor inputs."


allow things like
``` 
#!/usr/bin/env python
import torch

print("="*60)
print("Testing tensor slicing with torch.compile")
print("="*60)

# Test 1: Simple eager mode
print("\n1. Eager mode test:")
x = torch.randn(10)
idx = torch.tensor(4)
result = x[:idx]
print(f"   x[:idx] where idx=4: result.shape = {result.shape}")
assert result.shape[0] == 4
print("   ✓ Eager mode works!")

# Test 2: With torch.compile
print("\n2. Compiled mode test:")
def slice_fn(x, idx):
    return x[:idx]

try:
    compiled_fn = torch.compile(slice_fn)
    x = torch.randn(10)
    idx = torch.tensor(4)
    result = compiled_fn(x, idx)
    print(f"   Compiled x[:idx] where idx=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Compiled mode works!")
except Exception as e:
    print(f"   ✗ Compiled mode failed: {e}")
    import traceback
    traceback.print_exc()

# Test 3: With dynamic slicing from sum
print("\n3. Dynamic slicing with sum:")
def dynamic_slice_fn(x, lengths):
    idx = lengths.sum()
    return x[:idx]

try:
    compiled_fn = torch.compile(dynamic_slice_fn)
    x = torch.randn(10)
    lengths = torch.tensor([1, 1, 1, 1])
    result = compiled_fn(x, lengths)
    print(f"   Compiled x[:lengths.sum()] where sum=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Dynamic slicing works!")
except Exception as e:
    print(f"   ✗ Dynamic slicing failed: {e}")
    import traceback
    traceback.print_exc()

print("\n" + "="*60)
print("SUMMARY: Check results above")
print("="*60)

```


[ghstack-poisoned]

laithsakka committed:
Update on "[WIP] Support python slicing with tensor inputs."


allow things like
``` 
#!/usr/bin/env python
import torch

print("="*60)
print("Testing tensor slicing with torch.compile")
print("="*60)

# Test 1: Simple eager mode
print("\n1. Eager mode test:")
x = torch.randn(10)
idx = torch.tensor(4)
result = x[:idx]
print(f"   x[:idx] where idx=4: result.shape = {result.shape}")
assert result.shape[0] == 4
print("   ✓ Eager mode works!")

# Test 2: With torch.compile
print("\n2. Compiled mode test:")
def slice_fn(x, idx):
    return x[:idx]

try:
    compiled_fn = torch.compile(slice_fn)
    x = torch.randn(10)
    idx = torch.tensor(4)
    result = compiled_fn(x, idx)
    print(f"   Compiled x[:idx] where idx=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Compiled mode works!")
except Exception as e:
    print(f"   ✗ Compiled mode failed: {e}")
    import traceback
    traceback.print_exc()

# Test 3: With dynamic slicing from sum
print("\n3. Dynamic slicing with sum:")
def dynamic_slice_fn(x, lengths):
    idx = lengths.sum()
    return x[:idx]

try:
    compiled_fn = torch.compile(dynamic_slice_fn)
    x = torch.randn(10)
    lengths = torch.tensor([1, 1, 1, 1])
    result = compiled_fn(x, lengths)
    print(f"   Compiled x[:lengths.sum()] where sum=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Dynamic slicing works!")
except Exception as e:
    print(f"   ✗ Dynamic slicing failed: {e}")
    import traceback
    traceback.print_exc()

print("\n" + "="*60)
print("SUMMARY: Check results above")
print("="*60)

```


[ghstack-poisoned]

laithsakka committed:
Update on "[WIP] Support python slicing with tensor inputs."


allow things like
``` 
#!/usr/bin/env python
import torch

print("="*60)
print("Testing tensor slicing with torch.compile")
print("="*60)

# Test 1: Simple eager mode
print("\n1. Eager mode test:")
x = torch.randn(10)
idx = torch.tensor(4)
result = x[:idx]
print(f"   x[:idx] where idx=4: result.shape = {result.shape}")
assert result.shape[0] == 4
print("   ✓ Eager mode works!")

# Test 2: With torch.compile
print("\n2. Compiled mode test:")
def slice_fn(x, idx):
    return x[:idx]

try:
    compiled_fn = torch.compile(slice_fn)
    x = torch.randn(10)
    idx = torch.tensor(4)
    result = compiled_fn(x, idx)
    print(f"   Compiled x[:idx] where idx=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Compiled mode works!")
except Exception as e:
    print(f"   ✗ Compiled mode failed: {e}")
    import traceback
    traceback.print_exc()

# Test 3: With dynamic slicing from sum
print("\n3. Dynamic slicing with sum:")
def dynamic_slice_fn(x, lengths):
    idx = lengths.sum()
    return x[:idx]

try:
    compiled_fn = torch.compile(dynamic_slice_fn)
    x = torch.randn(10)
    lengths = torch.tensor([1, 1, 1, 1])
    result = compiled_fn(x, lengths)
    print(f"   Compiled x[:lengths.sum()] where sum=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Dynamic slicing works!")
except Exception as e:
    print(f"   ✗ Dynamic slicing failed: {e}")
    import traceback
    traceback.print_exc()

print("\n" + "="*60)
print("SUMMARY: Check results above")
print("="*60)

```


[ghstack-poisoned]

laithsakka committed:
Update on "[WIP] Support python slicing with tensor inputs."


allow things like
``` 
#!/usr/bin/env python
import torch

print("="*60)
print("Testing tensor slicing with torch.compile")
print("="*60)

# Test 1: Simple eager mode
print("\n1. Eager mode test:")
x = torch.randn(10)
idx = torch.tensor(4)
result = x[:idx]
print(f"   x[:idx] where idx=4: result.shape = {result.shape}")
assert result.shape[0] == 4
print("   ✓ Eager mode works!")

# Test 2: With torch.compile
print("\n2. Compiled mode test:")
def slice_fn(x, idx):
    return x[:idx]

try:
    compiled_fn = torch.compile(slice_fn)
    x = torch.randn(10)
    idx = torch.tensor(4)
    result = compiled_fn(x, idx)
    print(f"   Compiled x[:idx] where idx=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Compiled mode works!")
except Exception as e:
    print(f"   ✗ Compiled mode failed: {e}")
    import traceback
    traceback.print_exc()

# Test 3: With dynamic slicing from sum
print("\n3. Dynamic slicing with sum:")
def dynamic_slice_fn(x, lengths):
    idx = lengths.sum()
    return x[:idx]

try:
    compiled_fn = torch.compile(dynamic_slice_fn)
    x = torch.randn(10)
    lengths = torch.tensor([1, 1, 1, 1])
    result = compiled_fn(x, lengths)
    print(f"   Compiled x[:lengths.sum()] where sum=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Dynamic slicing works!")
except Exception as e:
    print(f"   ✗ Dynamic slicing failed: {e}")
    import traceback
    traceback.print_exc()

print("\n" + "="*60)
print("SUMMARY: Check results above")
print("="*60)

```


[ghstack-poisoned]

laithsakka commented:
great i just need to add unit tests!

laithsakka committed:
Update on "[WIP] Support python slicing with tensor inputs."


allow things like
``` 
#!/usr/bin/env python
import torch

print("="*60)
print("Testing tensor slicing with torch.compile")
print("="*60)

# Test 1: Simple eager mode
print("\n1. Eager mode test:")
x = torch.randn(10)
idx = torch.tensor(4)
result = x[:idx]
print(f"   x[:idx] where idx=4: result.shape = {result.shape}")
assert result.shape[0] == 4
print("   ✓ Eager mode works!")

# Test 2: With torch.compile
print("\n2. Compiled mode test:")
def slice_fn(x, idx):
    return x[:idx]

try:
    compiled_fn = torch.compile(slice_fn)
    x = torch.randn(10)
    idx = torch.tensor(4)
    result = compiled_fn(x, idx)
    print(f"   Compiled x[:idx] where idx=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Compiled mode works!")
except Exception as e:
    print(f"   ✗ Compiled mode failed: {e}")
    import traceback
    traceback.print_exc()

# Test 3: With dynamic slicing from sum
print("\n3. Dynamic slicing with sum:")
def dynamic_slice_fn(x, lengths):
    idx = lengths.sum()
    return x[:idx]

try:
    compiled_fn = torch.compile(dynamic_slice_fn)
    x = torch.randn(10)
    lengths = torch.tensor([1, 1, 1, 1])
    result = compiled_fn(x, lengths)
    print(f"   Compiled x[:lengths.sum()] where sum=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Dynamic slicing works!")
except Exception as e:
    print(f"   ✗ Dynamic slicing failed: {e}")
    import traceback
    traceback.print_exc()

print("\n" + "="*60)
print("SUMMARY: Check results above")
print("="*60)

```


[ghstack-poisoned]

laithsakka commented:
AI told me ....
idk if this is ok 

It is not always the case that calling .item() on a size-1 tensor preserves the semantics of slicing, because there is a crucial difference in how the resulting value is handled: its impact on the computational graph. 

The key difference: gradient tracking

When you use a size-1 tensor directly for slicing, the slicing operation itself is non-differentiable. Any upstream gradient information from the tensor used for the index is discarded, as you cannot differentiate with respect to a discrete index.

When you call.item()on the tensor first, you explicitly break the computational graph. The single value is extracted as a standard Python number, and any gradient information associated with that tensor is completely lost. 

For the purpose of the slice operation itself, the two methods are semantically equivalent—the same slice is performed. The difference lies in what happens if you later compute gradients on a loss function that depends on the result of the slicing operation.

laithsakka committed:
Update on "[WIP] Support python slicing with tensor inputs."


allow things like
``` 
#!/usr/bin/env python
import torch

print("="*60)
print("Testing tensor slicing with torch.compile")
print("="*60)

# Test 1: Simple eager mode
print("\n1. Eager mode test:")
x = torch.randn(10)
idx = torch.tensor(4)
result = x[:idx]
print(f"   x[:idx] where idx=4: result.shape = {result.shape}")
assert result.shape[0] == 4
print("   ✓ Eager mode works!")

# Test 2: With torch.compile
print("\n2. Compiled mode test:")
def slice_fn(x, idx):
    return x[:idx]

try:
    compiled_fn = torch.compile(slice_fn)
    x = torch.randn(10)
    idx = torch.tensor(4)
    result = compiled_fn(x, idx)
    print(f"   Compiled x[:idx] where idx=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Compiled mode works!")
except Exception as e:
    print(f"   ✗ Compiled mode failed: {e}")
    import traceback
    traceback.print_exc()

# Test 3: With dynamic slicing from sum
print("\n3. Dynamic slicing with sum:")
def dynamic_slice_fn(x, lengths):
    idx = lengths.sum()
    return x[:idx]

try:
    compiled_fn = torch.compile(dynamic_slice_fn)
    x = torch.randn(10)
    lengths = torch.tensor([1, 1, 1, 1])
    result = compiled_fn(x, lengths)
    print(f"   Compiled x[:lengths.sum()] where sum=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Dynamic slicing works!")
except Exception as e:
    print(f"   ✗ Dynamic slicing failed: {e}")
    import traceback
    traceback.print_exc()

print("\n" + "="*60)
print("SUMMARY: Check results above")
print("="*60)

```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "[WIP] Support python slicing with tensor inputs."


allow things like
``` 
#!/usr/bin/env python
import torch

print("="*60)
print("Testing tensor slicing with torch.compile")
print("="*60)

# Test 1: Simple eager mode
print("\n1. Eager mode test:")
x = torch.randn(10)
idx = torch.tensor(4)
result = x[:idx]
print(f"   x[:idx] where idx=4: result.shape = {result.shape}")
assert result.shape[0] == 4
print("   ✓ Eager mode works!")

# Test 2: With torch.compile
print("\n2. Compiled mode test:")
def slice_fn(x, idx):
    return x[:idx]

try:
    compiled_fn = torch.compile(slice_fn)
    x = torch.randn(10)
    idx = torch.tensor(4)
    result = compiled_fn(x, idx)
    print(f"   Compiled x[:idx] where idx=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Compiled mode works!")
except Exception as e:
    print(f"   ✗ Compiled mode failed: {e}")
    import traceback
    traceback.print_exc()

# Test 3: With dynamic slicing from sum
print("\n3. Dynamic slicing with sum:")
def dynamic_slice_fn(x, lengths):
    idx = lengths.sum()
    return x[:idx]

try:
    compiled_fn = torch.compile(dynamic_slice_fn)
    x = torch.randn(10)
    lengths = torch.tensor([1, 1, 1, 1])
    result = compiled_fn(x, lengths)
    print(f"   Compiled x[:lengths.sum()] where sum=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Dynamic slicing works!")
except Exception as e:
    print(f"   ✗ Dynamic slicing failed: {e}")
    import traceback
    traceback.print_exc()

print("\n" + "="*60)
print("SUMMARY: Check results above")
print("="*60)

```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "[WIP] Support python slicing with tensor inputs."


allow things like
``` 
#!/usr/bin/env python
import torch

print("="*60)
print("Testing tensor slicing with torch.compile")
print("="*60)

# Test 1: Simple eager mode
print("\n1. Eager mode test:")
x = torch.randn(10)
idx = torch.tensor(4)
result = x[:idx]
print(f"   x[:idx] where idx=4: result.shape = {result.shape}")
assert result.shape[0] == 4
print("   ✓ Eager mode works!")

# Test 2: With torch.compile
print("\n2. Compiled mode test:")
def slice_fn(x, idx):
    return x[:idx]

try:
    compiled_fn = torch.compile(slice_fn)
    x = torch.randn(10)
    idx = torch.tensor(4)
    result = compiled_fn(x, idx)
    print(f"   Compiled x[:idx] where idx=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Compiled mode works!")
except Exception as e:
    print(f"   ✗ Compiled mode failed: {e}")
    import traceback
    traceback.print_exc()

# Test 3: With dynamic slicing from sum
print("\n3. Dynamic slicing with sum:")
def dynamic_slice_fn(x, lengths):
    idx = lengths.sum()
    return x[:idx]

try:
    compiled_fn = torch.compile(dynamic_slice_fn)
    x = torch.randn(10)
    lengths = torch.tensor([1, 1, 1, 1])
    result = compiled_fn(x, lengths)
    print(f"   Compiled x[:lengths.sum()] where sum=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Dynamic slicing works!")
except Exception as e:
    print(f"   ✗ Dynamic slicing failed: {e}")
    import traceback
    traceback.print_exc()

print("\n" + "="*60)
print("SUMMARY: Check results above")
print("="*60)

```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "[WIP] Support python slicing with tensor inputs."


allow things like
``` 
#!/usr/bin/env python
import torch

print("="*60)
print("Testing tensor slicing with torch.compile")
print("="*60)

# Test 1: Simple eager mode
print("\n1. Eager mode test:")
x = torch.randn(10)
idx = torch.tensor(4)
result = x[:idx]
print(f"   x[:idx] where idx=4: result.shape = {result.shape}")
assert result.shape[0] == 4
print("   ✓ Eager mode works!")

# Test 2: With torch.compile
print("\n2. Compiled mode test:")
def slice_fn(x, idx):
    return x[:idx]

try:
    compiled_fn = torch.compile(slice_fn)
    x = torch.randn(10)
    idx = torch.tensor(4)
    result = compiled_fn(x, idx)
    print(f"   Compiled x[:idx] where idx=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Compiled mode works!")
except Exception as e:
    print(f"   ✗ Compiled mode failed: {e}")
    import traceback
    traceback.print_exc()

# Test 3: With dynamic slicing from sum
print("\n3. Dynamic slicing with sum:")
def dynamic_slice_fn(x, lengths):
    idx = lengths.sum()
    return x[:idx]

try:
    compiled_fn = torch.compile(dynamic_slice_fn)
    x = torch.randn(10)
    lengths = torch.tensor([1, 1, 1, 1])
    result = compiled_fn(x, lengths)
    print(f"   Compiled x[:lengths.sum()] where sum=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Dynamic slicing works!")
except Exception as e:
    print(f"   ✗ Dynamic slicing failed: {e}")
    import traceback
    traceback.print_exc()

print("\n" + "="*60)
print("SUMMARY: Check results above")
print("="*60)

```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "[WIP] Support python slicing with tensor inputs."


allow things like
``` 
#!/usr/bin/env python
import torch

print("="*60)
print("Testing tensor slicing with torch.compile")
print("="*60)

# Test 1: Simple eager mode
print("\n1. Eager mode test:")
x = torch.randn(10)
idx = torch.tensor(4)
result = x[:idx]
print(f"   x[:idx] where idx=4: result.shape = {result.shape}")
assert result.shape[0] == 4
print("   ✓ Eager mode works!")

# Test 2: With torch.compile
print("\n2. Compiled mode test:")
def slice_fn(x, idx):
    return x[:idx]

try:
    compiled_fn = torch.compile(slice_fn)
    x = torch.randn(10)
    idx = torch.tensor(4)
    result = compiled_fn(x, idx)
    print(f"   Compiled x[:idx] where idx=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Compiled mode works!")
except Exception as e:
    print(f"   ✗ Compiled mode failed: {e}")
    import traceback
    traceback.print_exc()

# Test 3: With dynamic slicing from sum
print("\n3. Dynamic slicing with sum:")
def dynamic_slice_fn(x, lengths):
    idx = lengths.sum()
    return x[:idx]

try:
    compiled_fn = torch.compile(dynamic_slice_fn)
    x = torch.randn(10)
    lengths = torch.tensor([1, 1, 1, 1])
    result = compiled_fn(x, lengths)
    print(f"   Compiled x[:lengths.sum()] where sum=4: result.shape = {result.shape}")
    assert result.shape[0] == 4
    print("   ✓ Dynamic slicing works!")
except Exception as e:
    print(f"   ✗ Dynamic slicing failed: {e}")
    import traceback
    traceback.print_exc()

print("\n" + "="*60)
print("SUMMARY: Check results above")
print("="*60)

```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

Lucaskabela commented on torch/_dynamo/variables/lists.py:
```diff
@@ -1399,18 +1399,27 @@ def __init__(self, items, **kwargs) -> None:
+            assert tx is not None, (
+                "tx is required when slice indices are TensorVariables"
+            )
+            assert stop.size is None or all(s == 1 for s in stop.size)
```
I think we are moving away from asserts in the variables code, let's avoid them here - see https://github.com/pytorch/pytorch/pull/165430 for reference

Lucaskabela reviewed (commented):
Generally this looks good, to me but needs some cleanup of asserts and I would like to see a test with dynamic=True to make sure the checks before `.item()` calls have reasonable behavior

Lucaskabela reviewed (changes requested):
See comments - few minor changes suggested but this looks good otherwise

laithsakka commented:
> Generally this looks good, to me but needs some cleanup of asserts and I would like to see a test with dynamic=True to make sure the checks before `.item()` calls have reasonable behavior

ah i see i will just remove those checks as you suggested probably size will be none with the input tensor is dynamic.

laithsakka committed:
Update on " Support python slicing with tensor inputs."


when the slice is tensor, we decompose it to .item() call and pass the unbacked symbol to the slice to avoid DDE. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on " Support python slicing with tensor inputs."


when the slice is tensor, we decompose it to .item() call and pass the unbacked symbol to the slice to avoid DDE. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

Lucaskabela reviewed (approved):
Thanks for addressing feedback and fixing the test failures - the dtensor failure seems unrelated to me, so approving
